File: src/main/csharp/SunliteSharp/Runtime/SLClass.cs
namespace SunliteSharp.Runtime;

using Type = SunliteSharp.Core.Type;

public record SLClass(
    string Name,
    Dictionary<string, SLClosureObj> Methods,
    Dictionary<string, SLField> FieldDefaults,
    Dictionary<string, SLField> StaticFields,
    Dictionary<string, Type> TypeParams,
    bool IsAbstract = false)
{
    public override string ToString()
    {
        return $"<class '{Name}'>";
    }
    
    public SLClass Copy()
    {
        return new SLClass(
            Name,
            Methods.ToDictionary(kv => kv.Key, kv => (SLClosureObj)kv.Value.Copy()),
            FieldDefaults.ToDictionary(kv => kv.Key, kv => kv.Value.Copy()),
            StaticFields.ToDictionary(kv => kv.Key, kv => kv.Value.Copy()),
            new Dictionary<string, Type>(TypeParams),
            IsAbstract
        );
    }
};

File: src/main/kotlin/sunsetsatellite/sunlite/vm/SLClass.kt
package sunsetsatellite.sunlite.vm

import sunsetsatellite.sunlite.lang.Type

class SLClass(
    val name: String,
    val methods: MutableMap<String, SLClosureObj>,
    val fieldDefaults: MutableMap<String, SLField>,
    val staticFields: MutableMap<String, SLField>,
    val typeParams: MutableMap<String, Type>,
    val isAbstract: Boolean = false
) {

    override fun toString(): String {
        return "<class '${name}'>"
    }

    fun copy(): SLClass {
        return SLClass(
            name,
            methods.mapValues { it.value.copy() as SLClosureObj }.toMutableMap(),
            fieldDefaults.mapValues { it.value.copy() }.toMutableMap(),
            staticFields.mapValues { it.value.copy() }.toMutableMap(),
            typeParams.toMutableMap(),
            isAbstract
        )
    }
}